docs/source/customizations/overview.mdx

```diff
@@ -11,4 +11,142 @@ TODO:
+
+The trait itself provides default implementations of each hook that returns the service as is.
+
+Each plugin may have configuration. If specified, plugin config will automatically loaded and passed to the `new` method on initialization. You'll be able to configure your plugin via the Router configuration yaml.
+
+```rust=
+#[async_trait]
+pub trait Plugin: Send + Sync + 'static + Sized {
+    type Config: JsonSchema + DeserializeOwned;
+
+    fn new(config: Self::Config) -> Result<Self, BoxError>;
+
+    async fn startup(&mut self) -> Result<(), BoxError> {
+        Ok(())
+    }
+    async fn shutdown(&mut self) -> Result<(), BoxError> {
+        Ok(())
+    }
+
+    fn router_service(
+        &mut self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        service
+    }
+
+    fn query_planning_service(
+        &mut self,
+        service: BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError>,
+    ) -> BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError> {
+        service
+    }
+
+    fn execution_service(
+        &mut self,
+        service: BoxService<ExecutionRequest, ExecutionResponse, BoxError>,
+    ) -> BoxService<ExecutionRequest, ExecutionResponse, BoxError> {
+        service
+    }
+
+    fn subgraph_service(
+        &mut self,
+        _name: &str,
+        service: BoxService<SubgraphRequest, SubgraphResponse, BoxError>,
+    ) -> BoxService<SubgraphRequest, SubgraphResponse, BoxError> {
+        service
+    }
+
+    fn name(&self) -> &'static str {
+        get_type_of(self)
+    }
+}
+```
+
+### `ServiceBuilder`
+
+A plugin may augment any of the services to provide extra functionality.
+ServiceBuilder provides common building blocks that removes much of the complexity of writing a plugin. These are called layers.
+
+```rust=
+    fn router_service(
+        &mut self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        // Always use service builder to compose your plugins.
+        // It provides off the shelf building blocks for your plugin.
+        ServiceBuilder::new()
+            // Some example service builder methods:
+            // .map_request()
+            // .map_response()
+            // .rate_limit()
+            // .checkpoint()
+            // .timeout()
+            .service(service)
+            .boxed()
+    }
+```
+
+Tower-rs, on which the router is based comes with many off the shelf layers. In addition Apollo provide layers that cover common functionality and integration with thrid party products.
+
+### Registering your plugin
+
+To enable the Router to discover your plugin it must be registered. Plugin configuration can then be added to the Router configuration yaml to activate your plugin.  
+
+```rust=
+register_plugin!(
+    "example",
+    "my_custom_plugin",
+    MyPlugin
```

Comment:
Maybe add some text to explain that the first string is "group name". The second string is "plugin name". When composed they are "group name"."plugin.name".